pkg/controller/replica.go

```diff
@@ -0,0 +1,338 @@
+	}
+
+	skipSet := map[string]bool{}
+	for _, replicaNode := range replicaNodes {
+		skipSet[replicaNode] = true
+	}
+
+	filtered := []string{}
+	numFiltered := 0
+	for _, node := range nodes {
+		if numFiltered >= numReplica {
+			break
+		}
+		if skipSet[node] {
+			continue
+		}
+		filtered = append(filtered, node)
+		numFiltered++
+	}
+
+	return filtered, unlocks, nil
+}
+
+func (r *ReconcileReplica) createReplicas(ctx context.Context, numReplica int, volumeName, volumeID string) error {
+	// if volume is scheduled for clean up, skip replica creation
+	if _, cleanUpScheduled := r.cleanUpMap.Load(volumeName); cleanUpScheduled {
+		return nil
+	}
+
+	nodes, unlocks, err := r.getNodesForReplica(ctx, volumeName, numReplica)
+	defer unlockPodLocks(unlocks)
+	if err != nil {
+		klog.Errorf("failed to get a list of nodes for replica attachment: %v", err)
+		return err
+	}
+
+	for _, node := range nodes {
+		if err := createReplica(ctx, r, volumeID, node); err != nil {
+			klog.Errorf("failed to create replica azVolumeAttachment for volume %s: %v", volumeName, err)
+			return err
+		}
+	}
+	return nil
+}
+
+func unlockPodLocks(unlocks []func()) {
+	for _, unlock := range unlocks {
+		unlock()
+	}
+}
+
+func NewReplicaController(mgr manager.Manager, azVolumeClient azClientSet.Interface, kubeClient kubeClientSet.Interface, controllerSharedState *SharedState) (*ReconcileReplica, error) {
+	logger := mgr.GetLogger().WithValues("controller", "replica")
+	reconciler := ReconcileReplica{
+		client:                mgr.GetClient(),
+		azVolumeClient:        azVolumeClient,
+		kubeClient:            kubeClient,
+		controllerSharedState: controllerSharedState,
+		deletionMap:           sync.Map{},
+		cleanUpMap:            sync.Map{},
+		mutexLocks:            sync.Map{},
+	}
+	c, err := controller.New("replica-controller", mgr, controller.Options{
+		MaxConcurrentReconciles: 10,
+		Reconciler:              &reconciler,
+		Log:                     logger,
+	})
+
+	if err != nil {
+		klog.Errorf("Failed to create replica controller. Error: (%v)", err)
+		return nil, err
+	}
+
+	klog.V(2).Info("Starting to watch AzVolumeAttachments.")
+
+	err = c.Watch(&source.Kind{Type: &v1alpha1.AzVolumeAttachment{}}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			azVolumeAttachment, ok := e.Object.(*v1alpha1.AzVolumeAttachment)
+			if ok && azVolumeAttachment.Spec.RequestedRole == v1alpha1.PrimaryRole {
+				return true
+			}
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// enqueue AzVolumeAttachment promotion events
+			old, oldOk := e.ObjectOld.(*v1alpha1.AzVolumeAttachment)
+			new, newOk := e.ObjectNew.(*v1alpha1.AzVolumeAttachment)
+			if oldOk && newOk && old.Spec.RequestedRole != new.Spec.RequestedRole {
+				return true
+			}
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
```

Comment:
Shouldn't we also listen to the delete events for the Primary AzVolumeAttachment objects as we are checking this scenario in the reconcile function above?